Mathlib/NumberTheory/Padics/PadicNumbers.lean

```diff
@@ -969,27 +959,39 @@ theorem norm_eq_pow_val {x : ℚ_[p]} : x ≠ 0 → ‖x‖ = (p : ℝ) ^ (-x.va
     simpa using hf'
 
 @[simp]
-theorem valuation_p : valuation (p : ℚ_[p]) = 1 := by
-  have h : (1 : ℝ) < p := mod_cast (Fact.out : p.Prime).one_lt
-  refine neg_injective ((zpow_right_strictMono₀ h).injective <| (norm_eq_pow_val ?_).symm.trans ?_)
-  · exact mod_cast (Fact.out : p.Prime).ne_zero
-  · simp
+lemma valuation_ratCast (q : ℚ) : valuation (q : ℚ_[p]) = padicValRat p q := by
+  rcases eq_or_ne q 0 with rfl | hq
+  · simp only [Rat.cast_zero, valuation_zero, padicValRat.zero]
+  refine neg_injective ((zpow_right_strictMono₀ (mod_cast hp.out.one_lt)).injective
+    <| (norm_eq_pow_val (mod_cast hq)).symm.trans ?_)
+  rw [padicNormE.eq_padicNorm, ← Rat.cast_natCast, ← Rat.cast_zpow, Rat.cast_inj]
+  exact padicNorm.eq_zpow_of_nonzero hq
+
+@[simp]
+lemma valuation_intCast (n : ℤ) : valuation (n : ℚ_[p]) = padicValInt p n := by
+  rw [← Rat.cast_intCast, valuation_ratCast, padicValRat.of_int]
+
+@[simp]
+lemma valuation_natCast (n : ℕ) : valuation (n : ℚ_[p]) = padicValNat p n := by
+  rw [← Rat.cast_natCast, valuation_ratCast, padicValRat.of_nat]
+
+@[simp]
+lemma valuation_one : valuation (1 : ℚ_[p]) = 0 := by
+  rw [← Nat.cast_one, valuation_natCast, padicValNat.one, cast_zero]
```

Comment:
Please also add `valuation_ofNat` with `OfNat.ofNat` and use it for `valuation_one`. (Don't forget to cargo-cult the `no_index` stuff.)

Reply:
What is this about `no_index`? I've not heard of it before.